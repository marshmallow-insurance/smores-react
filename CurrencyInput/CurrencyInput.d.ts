import React, { FocusEvent, FormEvent } from 'react';
import { CommonFieldProps } from '../fields/commonFieldTypes';
export interface Props extends CommonFieldProps {
    placeholder: string;
    name?: string;
    value: string;
    decimalNumber?: boolean;
    onBlur?: (e: FocusEvent<HTMLInputElement>) => void;
    min?: number;
    max?: number;
}
/** on change or on input required */
type InputProps = {
    /** on change is required and on input optional */
    onChange: (e: string) => void;
    onInputChange?: (e: FormEvent<HTMLInputElement>) => void;
} | {
    /** on input is required and on change optional */
    onChange?: (e: string) => void;
    onInputChange: (e: FormEvent<HTMLInputElement>) => void;
};
export type CurrencyInputProps = Props & InputProps;
export declare const CurrencyInput: React.ForwardRefExoticComponent<CurrencyInputProps & React.RefAttributes<HTMLInputElement>>;
export {};
