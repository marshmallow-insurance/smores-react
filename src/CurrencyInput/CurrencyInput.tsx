import React, {
  FocusEvent,
  FormEvent,
  forwardRef,
  ForwardedRef,
} from 'react'

import { Box } from '../Box'
import { Field } from '../fields/Field'
import { CommonFieldProps } from '../fields/commonFieldTypes'

import { useUniqueId } from '../utils/id'
import {
  Input,
  StyledFrontIcon,
} from '../fields/components/CommonInput'

export interface Props extends CommonFieldProps {
  placeholder: string
  name?: string
  value: string
  onBlur?: (e: FocusEvent<HTMLInputElement>) => void
  min?: number
  max?: number
}

/** on change or on input required */
type InputProps =
  | {
      /** on change is required and on input optional */
      onChange: (e: string) => void
      onInputChange?: (e: FormEvent<HTMLInputElement>) => void
    }
  | {
      /** on input is required and on change optional */
      onChange?: (e: string) => void
      onInputChange: (e: FormEvent<HTMLInputElement>) => void
    }

export type CurrencyInputProps = Props & InputProps

export const CurrencyInput = forwardRef(function CurrencyInput(
  {
    id: idProp,
    placeholder,
    name,
    value,
    onChange,
    onInputChange,
    onBlur,
    min,
    max,
    disabled = false,
    error = false,
    trailingIcon,
    fallbackStyle,
    ...fieldProps
  }: CurrencyInputProps,
  ref: ForwardedRef<HTMLInputElement>,
) {
  const id = useUniqueId(idProp)
  const frontIcon = "pound-regular"

  const formatCurrency = (value: string) => {
    const decimalIndex = value.indexOf('.')
    if (decimalIndex >= 0 && value.length > decimalIndex + 1) {
      const fractionalString = value.substring(decimalIndex + 1).substring(0, 2)
      return `${value.substring(0, decimalIndex)}.${fractionalString}`
    } else {
      return value
    }
  }

  const applyMinMax = (value: string) => {
    const numberValue = Number(value)
    if(min && numberValue < min) return min.toString()
    if(max && numberValue > max) return max.toString()
    return value
  }

  const handleChange = (value: string) => {
    if (onChange === undefined) return
    const EMPTY_INPUT = ''

    if (value === EMPTY_INPUT) {
      onChange(value)
    } else {
      const amount = formatCurrency(value)
      const normalisedValue = applyMinMax(amount)

      onChange(normalisedValue)
    }
  }

  return (
    <Field {...fieldProps} htmlFor={id} error={error}>
      <Box flex alignItems="center" justifyContent="flex-start">
        <StyledFrontIcon
          disabled={disabled}
          render={frontIcon}
          color="sesame"
        />
        <Input
          ref={ref}
          error={error}
          disabled={disabled}
          type="number"
          id={id}
          name={name}
          placeholder={placeholder}
          value={value}
          frontIcon={frontIcon}
          fallbackStyle={fallbackStyle}
          onChange={(e: FormEvent<HTMLInputElement>) => {
            onChange && handleChange(e.currentTarget.value)
            onInputChange && onInputChange(e)
          }}
          onBlur={onBlur}
        />
      </Box>
    </Field>
  )
})
