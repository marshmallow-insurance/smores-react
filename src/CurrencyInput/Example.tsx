import React, { useState } from 'react'

import { CurrencyInput } from './CurrencyInput'

export const Example = () => {
  const [value, setValue] = useState('')

  return (
    <form>
      <CurrencyInput
        id="numberInput"
        label="Amount paid"
        name="numberInput"
        onChange={(e) => setValue(e)}
        placeholder="100.00"
        value={value}
        min={-200}
        max={200}
        required
      />
    </form>
  )
}
