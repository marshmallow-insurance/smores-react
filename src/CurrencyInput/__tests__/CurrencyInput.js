import React from 'react';
import {render} from '@testing-library/react';
import 'jest-styled-components';

import {CurrencyInput} from '../CurrencyInput';
import { CurrencyInput } from 'CurrencyInput/CurrencyInput';

test('renders - number', () => {
  const {container} = render(
    <CurrencyInput
      type="number"
      id="CurrencyInput"
      label="Number"
      name="CurrencyInput"
      value=""
      onChange={str => {}}
      placeholder="0.00"
      required={true}
    />,
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('renders - error', () => {
  const {container} = render(
    <CurrencyInput
      type="number"
      id="currencyInput"
      label="Currency"
      name="currencyInput"
      value="200"
      onChange={str => {}}
      placeholder="100.00"
      prefix="$"
      strict={true}
      min={0}
      max={100}
      step={10}
      roundCurrency={false}
      error={true}
      errorMsg="Oh gosh, this is an error isn't it!"
    />,
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('renders - disabled', () => {
  const {container} = render(
    <CurrencyInput
      type="number"
      id="currencyInput"
      label="Currency"
      name="currencyInput"
      value=""
      onChange={str => {}}
      placeholder="100.00"
      disabled={true}
    />,
  );
  expect(container.firstChild).toMatchSnapshot();
});
