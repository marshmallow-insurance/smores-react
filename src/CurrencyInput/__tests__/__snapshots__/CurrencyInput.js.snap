// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders - disabled 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c2 {
  font-size: 10px;
  line-height: 12px;
  font-weight: 500;
  text-transform: uppercase;
  text-align: left;
  cursor: inherit;
  color: #636768;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 100%;
  border: 0;
}

.c6 {
  border: none;
  color: #292924;
  font-size: 16px;
  width: 100%;
  outline: none;
  cursor: not-allowed;
  opacity: 0.5;
  padding: 18px 14px;
  background-color: #FFFFFF;
  border: 2px solid #DAD2C4;
  border-radius: 12px;
  height: auto;
  padding-left: 42px;
  -moz-appearance: textfield;
}

.c6::-webkit-outer-spin-button,
.c6::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c6::-webkit-input-placeholder {
  color: #636768;
}

.c6::-moz-placeholder {
  color: #636768;
}

.c6:-ms-input-placeholder {
  color: #636768;
}

.c6::placeholder {
  color: #636768;
}

.c5 {
  position: relative;
  left: 36px;
  margin-left: -24px;
  z-index: 1;
  opacity: 0.5;
  color: #292924;
}

@media (min-width:0px) {
  .c1 {
    margin-bottom: 4px;
  }
}

@media (min-width:0px) {
  .c3 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
  }
}

@media (min-width:0px) {
  .c3 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<div
  class="c0"
  type="number"
>
  <label
    class="c1 c2"
    color="sesame"
    cursor="inherit"
    for="currencyInput"
    title=""
  >
    Currency
  </label>
  <div
    class=""
  >
    <div
      class="c3"
    >
      <span
        class="c4 c5"
        color="sesame"
        disabled=""
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.99 15.876l1.669.168c0 .685-.154 1.322-.462 1.91-.3.587-.77 1.077-1.406 1.469-.637.385-1.382.577-2.235.577H6v-1.637c.784-.392 1.43-.983 1.941-1.773.518-.791.777-1.617.777-2.477 0-.469-.07-.983-.21-1.543H6.063v-1.616h1.91a45.567 45.567 0 01-.388-1.102 8.872 8.872 0 01-.263-1.038 6.013 6.013 0 01-.126-1.176c0-.874.224-1.668.672-2.382a4.6 4.6 0 011.794-1.658A5.12 5.12 0 0112.107 3c1.456 0 2.6.378 3.432 1.133.832.756 1.273 1.735 1.322 2.939l-1.69.293a3.215 3.215 0 00-.178-.955 2.874 2.874 0 00-.472-.87c-.21-.28-.518-.5-.924-.662-.405-.168-.884-.251-1.437-.251-.889 0-1.637.293-2.246.881-.601.588-.902 1.319-.902 2.193 0 .126.007.26.02.399l.043.367c.02.105.052.245.094.42.05.175.084.304.105.388.021.077.067.22.136.43.07.21.112.343.126.4.021.048.074.195.158.44.084.238.133.374.147.41h4.712v1.615h-4.209c.133.511.2.994.2 1.448 0 1.784-.71 3.233-2.13 4.345h4.91c.518 0 .963-.077 1.333-.23.371-.162.648-.375.83-.641.181-.266.31-.528.388-.787.077-.266.115-.542.115-.83z"
            fill="#636768"
          />
        </svg>
      </span>
      <input
        class="c6"
        disabled=""
        id="currencyInput"
        name="currencyInput"
        placeholder="100.00"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`renders - error 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c2 {
  font-size: 10px;
  line-height: 12px;
  font-weight: 500;
  text-transform: uppercase;
  text-align: left;
  cursor: inherit;
  color: #DF2929;
}

.c8 {
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  text-align: left;
  cursor: inherit;
  color: #DF2929;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 100%;
  border: 0;
}

.c6 {
  border: none;
  color: #DF2929;
  font-size: 16px;
  width: 100%;
  outline: none;
  cursor: initial;
  opacity: 1;
  padding: 18px 14px;
  background-color: #FFFFFF;
  border: 2px solid #DF2929;
  border-radius: 12px;
  height: auto;
  padding-left: 42px;
  -moz-appearance: textfield;
}

.c6::-webkit-outer-spin-button,
.c6::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c6::-webkit-input-placeholder {
  color: #636768;
}

.c6::-moz-placeholder {
  color: #636768;
}

.c6:-ms-input-placeholder {
  color: #636768;
}

.c6::placeholder {
  color: #636768;
}

.c5 {
  position: relative;
  left: 36px;
  margin-left: -24px;
  z-index: 1;
  opacity: 1;
  color: #292924;
}

@media (min-width:0px) {
  .c1 {
    margin-bottom: 4px;
  }
}

@media (min-width:0px) {
  .c3 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
  }
}

@media (min-width:0px) {
  .c3 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

@media (min-width:0px) {
  .c7 {
    margin-top: 8px;
  }
}

<div
  class="c0"
  type="number"
>
  <label
    class="c1 c2"
    color="strawberry"
    cursor="inherit"
    for="currencyInput"
    title=""
  >
    Currency
  </label>
  <div
    class=""
  >
    <div
      class="c3"
    >
      <span
        class="c4 c5"
        color="sesame"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.99 15.876l1.669.168c0 .685-.154 1.322-.462 1.91-.3.587-.77 1.077-1.406 1.469-.637.385-1.382.577-2.235.577H6v-1.637c.784-.392 1.43-.983 1.941-1.773.518-.791.777-1.617.777-2.477 0-.469-.07-.983-.21-1.543H6.063v-1.616h1.91a45.567 45.567 0 01-.388-1.102 8.872 8.872 0 01-.263-1.038 6.013 6.013 0 01-.126-1.176c0-.874.224-1.668.672-2.382a4.6 4.6 0 011.794-1.658A5.12 5.12 0 0112.107 3c1.456 0 2.6.378 3.432 1.133.832.756 1.273 1.735 1.322 2.939l-1.69.293a3.215 3.215 0 00-.178-.955 2.874 2.874 0 00-.472-.87c-.21-.28-.518-.5-.924-.662-.405-.168-.884-.251-1.437-.251-.889 0-1.637.293-2.246.881-.601.588-.902 1.319-.902 2.193 0 .126.007.26.02.399l.043.367c.02.105.052.245.094.42.05.175.084.304.105.388.021.077.067.22.136.43.07.21.112.343.126.4.021.048.074.195.158.44.084.238.133.374.147.41h4.712v1.615h-4.209c.133.511.2.994.2 1.448 0 1.784-.71 3.233-2.13 4.345h4.91c.518 0 .963-.077 1.333-.23.371-.162.648-.375.83-.641.181-.266.31-.528.388-.787.077-.266.115-.542.115-.83z"
            fill="#636768"
          />
        </svg>
      </span>
      <input
        class="c6"
        id="currencyInput"
        name="currencyInput"
        placeholder="100.00"
        type="number"
        value="200"
      />
    </div>
  </div>
  <span
    class="c7 c8"
    color="strawberry"
    cursor="inherit"
    title=""
  >
    Oh gosh, this is an error isn't it!
  </span>
</div>
`;

exports[`renders - number 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c2 {
  font-size: 10px;
  line-height: 12px;
  font-weight: 500;
  text-transform: uppercase;
  text-align: left;
  cursor: inherit;
  color: #636768;
}

.c3 {
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
  text-align: left;
  cursor: inherit;
  color: #DF2929;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 100%;
  border: 0;
}

.c7 {
  border: none;
  color: #292924;
  font-size: 16px;
  width: 100%;
  outline: none;
  cursor: initial;
  opacity: 1;
  padding: 18px 14px;
  background-color: #FFFFFF;
  border: 2px solid #DAD2C4;
  border-radius: 12px;
  height: auto;
  padding-left: 42px;
  -moz-appearance: textfield;
}

.c7:hover,
.c7:focus-within {
  border-color: #A1957C;
}

.c7::-webkit-outer-spin-button,
.c7::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c7::-webkit-input-placeholder {
  color: #636768;
}

.c7::-moz-placeholder {
  color: #636768;
}

.c7:-ms-input-placeholder {
  color: #636768;
}

.c7::placeholder {
  color: #636768;
}

.c6 {
  position: relative;
  left: 36px;
  margin-left: -24px;
  z-index: 1;
  opacity: 1;
  color: #292924;
}

@media (min-width:0px) {
  .c1 {
    margin-bottom: 4px;
  }
}

@media (min-width:0px) {
  .c4 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
  }
}

@media (min-width:0px) {
  .c4 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<div
  class="c0"
  type="number"
>
  <label
    class="c1 c2"
    color="sesame"
    cursor="inherit"
    for="CurrencyInput"
    title=""
  >
    Number
    <span
      class="c3"
      color="strawberry"
      cursor="inherit"
      title=""
    >
      *
    </span>
  </label>
  <div
    class=""
  >
    <div
      class="c4"
    >
      <span
        class="c5 c6"
        color="sesame"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.99 15.876l1.669.168c0 .685-.154 1.322-.462 1.91-.3.587-.77 1.077-1.406 1.469-.637.385-1.382.577-2.235.577H6v-1.637c.784-.392 1.43-.983 1.941-1.773.518-.791.777-1.617.777-2.477 0-.469-.07-.983-.21-1.543H6.063v-1.616h1.91a45.567 45.567 0 01-.388-1.102 8.872 8.872 0 01-.263-1.038 6.013 6.013 0 01-.126-1.176c0-.874.224-1.668.672-2.382a4.6 4.6 0 011.794-1.658A5.12 5.12 0 0112.107 3c1.456 0 2.6.378 3.432 1.133.832.756 1.273 1.735 1.322 2.939l-1.69.293a3.215 3.215 0 00-.178-.955 2.874 2.874 0 00-.472-.87c-.21-.28-.518-.5-.924-.662-.405-.168-.884-.251-1.437-.251-.889 0-1.637.293-2.246.881-.601.588-.902 1.319-.902 2.193 0 .126.007.26.02.399l.043.367c.02.105.052.245.094.42.05.175.084.304.105.388.021.077.067.22.136.43.07.21.112.343.126.4.021.048.074.195.158.44.084.238.133.374.147.41h4.712v1.615h-4.209c.133.511.2.994.2 1.448 0 1.784-.71 3.233-2.13 4.345h4.91c.518 0 .963-.077 1.333-.23.371-.162.648-.375.83-.641.181-.266.31-.528.388-.787.077-.266.115-.542.115-.83z"
            fill="#636768"
          />
        </svg>
      </span>
      <input
        class="c7"
        id="CurrencyInput"
        name="CurrencyInput"
        placeholder="0.00"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;
